nemo_skills/inference/generate.py

```diff
@@ -511,49 +431,55 @@ def prefill_generation(self, data_point) -> dict | None:
-                    self.dump_outputs([prefill_output], [data_point], fout)
-                else:
-                    data_points_batch.append(data_point)
-
-                if len(data_points_batch) == self.cfg.max_concurrent_requests or idx == len(data) - 1:
-
-                    for data_point in data_points_batch:
-                        # registering current time to calculate total generation time
-                        data_point['generation_start_time'] = time.time()
-
-                    if self.cfg.multi_turn_key is None:
-                        outputs = self.llm_generate(data_points_batch, data)
-                    else:
-                        outputs = self.llm_generate_multi_turn(data_points_batch, data)
-                    self.dump_outputs(outputs, data_points_batch, fout)
-                    data_points_batch = []
-
-    def get_llm_generations(self, requests_in_progress, generations):
-        """Get the completed LLM generations that were submitted asynchronously."""
+    async def process_single_datapoint(self, data_point, all_data):
+        generation_params = {
+            "prompts": [self.fill_prompt(data_point, all_data)],
+            "stop_phrases": combine_stop_phrases(
+                self.prompt.stop_phrases if self.prompt is not None else None, self.extra_stop_phrases
+            ),
+            **asdict(self.cfg.inference),
+            **self.extra_generate_params,
+        }
 
-        gen_ids = list(requests_in_progress.values())
-        outputs = self.llm.get_generations(gen_ids)
+        if self.cfg.code_execution:
+            if self.cfg.override_max_code_executions and self.cfg.total_code_executions_in_prompt is not None:
+                max_code_executions_values = [data_point['total_code_executions']]
+                generation_params['max_code_executions'] = max_code_executions_values
 
-        for dp_idx, output in zip(requests_in_progress.keys(), outputs):
-            generations[dp_idx] = output
+        return await self.llm.generate_asyncio(**generation_params)
 
-        return requests_in_progress, generations
 
-    def async_loop(self, data):
-        """Async loop to generate generations."""
+    async def _process_single_datapoint_with_semaphore(self, data_point, all_data, fout, pbar):
+        """Process a single data point with semaphore control."""
+        async with self.semaphore:
+            # registering current time to calculate total generation time
+            data_point['generation_start_time'] = time.time()
+            
+            try:
+                # Generate output for this single data point
+                output = await self.process_single_datapoint(data_point, all_data)
+                
+                # Thread-safe output writing
+                async with self.output_lock:
+                    self.dump_outputs([output], [data_point], fout)
+                    pbar.update(1)
+                    
+            except Exception as e:
```

Comment:
does the exception not get printed otherwise? If it's already part of the stacktrace, do we need another LOG.error or maybe just let it fail without try-except?